module/bdev/lvol/vbdev_lvol.c

```diff
@@ -119,6 +149,87 @@ _vbdev_lvol_change_bdev_alias(struct spdk_lvol *lvol, const char *new_lvol_name)
 	return 0;
 }
 
+static void
+vbdev_lvol_channel_destroy_resource(struct spdk_lvol_channel *ch)
+{
+	struct freeze_range *range;
+
+	while (!TAILQ_EMPTY(&ch->freezed_ranges)) {
+		range = TAILQ_FIRST(&ch->freezed_ranges);
+		TAILQ_REMOVE(&ch->freezed_ranges, range, tailq);
+		free(range);
+	}
+
+	spdk_put_io_channel(ch->bs_channel);
+
+	assert(TAILQ_EMPTY(&ch->queued_io));
+	assert(TAILQ_EMPTY(&ch->submitted_io));
+}
+
+static int
+vbdev_lvol_channel_create(void *io_device, void *ctx_buf)
+{
+	struct spdk_lvol		*lvol = io_device;
+	struct spdk_lvol_channel	*channel = ctx_buf;
+	struct freeze_range		*range;
+
+	channel->bs_channel = spdk_lvol_get_io_channel(lvol);
+	TAILQ_INIT(&channel->submitted_io);
+	TAILQ_INIT(&channel->queued_io);
+	TAILQ_INIT(&channel->freezed_ranges);
+
+	spdk_spin_lock(&lvol->spinlock);
+	TAILQ_FOREACH(range, &lvol->freezed_ranges, tailq) {
+		struct freeze_range *new_range;
+
+		new_range = calloc(1, sizeof(*new_range));
+		if (new_range == NULL) {
+			spdk_spin_unlock(&lvol->spinlock);
+			vbdev_lvol_channel_destroy_resource(channel);
+			return -1;
+		}
+		new_range->offset = range->offset;
+		new_range->length = range->length;
+		new_range->freezed_ctx = range->freezed_ctx;
+		TAILQ_INSERT_TAIL(&channel->freezed_ranges, new_range, tailq);
+	}
+	spdk_spin_unlock(&lvol->spinlock);
+
+	return 0;
+}
+
+static void
+vbdev_lvol_channel_destroy(void *io_device, void *ctx_buf)
+{
+	struct spdk_lvol_channel *channel = ctx_buf;
+
+	vbdev_lvol_channel_destroy_resource(channel);
+}
+
+static int
+vbdev_lvol_register(struct spdk_lvol *lvol, struct spdk_bdev *lvol_bdev)
+{
+	int rc;
+
+	/* We must register lvol as io_device before to register lvol bdev */
+	spdk_io_device_register(lvol, vbdev_lvol_channel_create, vbdev_lvol_channel_destroy,
+				sizeof(struct spdk_lvol_channel), "lvol");
+
+	rc = spdk_bdev_register(lvol_bdev);
+	if (rc) {
+		spdk_io_device_unregister(lvol, NULL);
+	}
+
+	return rc;
+}
+
+static void
+vbdev_lvol_unregister(struct spdk_lvol *lvol, spdk_bdev_unregister_cb cb_fn, void *cb_arg)
+{
+	spdk_io_device_unregister(lvol, NULL);
```

Comment:
Shall we check `lvol`
```
	if (lvol == NULL) {
		return
	}
```

Reply:
The function `spdk_io_device_unregister` calls `io_device_get`, which return NULL if the lvol is not found(and it will be not found if lvol si NULL), and so we will break on an assert. 
However you are right, to put a check is fine.